### PR TITLE
Rework storage. Give node creation responsibility to factory function

### PR DESCRIPTION
**Pull request by denys-hrulov**
-Storage now only owns tasks created outside it. Id generator now in service. 
-Added enum class representing storage insertion/deletion result.
-These changes forced to change several methods in TaskService somehow.

### CODE REVIEW

File: src/memory_model/TaskStrorageInterface.h [hunk 1]
```diff
@@ -10,10 +10,23 @@
 class TaskStrorageInterface {
 
 public:
-    virtual std::shared_ptr<TaskNode>   createTask(const Task&) = 0;
-    virtual void                        eraseTask(TaskID id_erase) = 0;
+    class Result {
+
+    public:
+        enum class onAdd {
```
Comment (vladimir-moiseiev) [inline src/memory_model/TaskStrorageInterface.h]: This is a total waste. How will you use it?

**Review by vladimir-moiseiev — CHANGES_REQUESTED** [review]

File: src/memory_model/TaskStrorageInterface.h (hunk 1 above)
Comment (denys-hrulov) [reply src/memory_model/TaskStrorageInterface.h]: I do want to mix up results from different operations. Maybe i found poor solution.

Comment (vladimir-moiseiev) [reply src/memory_model/TaskStrorageInterface.h]: And why do you need that?

Comment (denys-hrulov) [reply src/memory_model/TaskStrorageInterface.h]: My idea was to narrow possibilities of error returning value and exclude impossible.
TASK_NOT_FOND does not make any sense when adding task.

SO, dividing into different classes can improve IDE support and allow consider less possibilities of return value.

Comment (vladimir-moiseiev) [reply src/memory_model/TaskStrorageInterface.h]: > My idea was to narrow possibilities of error returning value

Why do you need that?
> and exclude impossible.

what impossible?

> SO, dividing into different classes can improve IDE support and allow consider less possibilities of return value.

Do you need IDE support? What support are you talking about?

Comment (denhrulyov) [reply src/memory_model/TaskStrorageInterface.h:16]: > what impossible?
For example, 
TASK_NOT_FOUND is impossible when adding task to storage.
> Do you need IDE support? What support are you talking about?
IDE will show only possible variants of result.

Comment (denys-hrulov) [reply src/memory_model/TaskStrorageInterface.h]: > what impossible?

For example,
TASK_NOT_FOUND is impossible when adding task to storage.

> Do you need IDE support? What support are you talking about?

IDE will show only possible variants of result.

Comment (denys-hrulov) [reply src/memory_model/TaskStrorageInterface.h]: I just want to explain the idea. no problem to remove the feature

Comment (vladimir-moiseiev) [reply src/memory_model/TaskStrorageInterface.h]: It doesn't need to be a part of the same class `Result `. Mixing everything in a single type is a bad solution. You can reuse values, like `Task_not_Found`, but you don't need a hierarchy.

Also `onAdd` name is bad. This is the style for event handlers, not results.

Comment (vladimir-moiseiev) [reply src/memory_model/TaskStrorageInterface.h]: Also, this is not a public api, this is a working class. Why do you need results that complex?
your public API in taskService lacks that

Comment (denys-hrulov) [reply src/memory_model/TaskStrorageInterface.h]: So it's still better to have a single enum for storage results, or even bool?

**Pull request merged**